core/src/io/anuke/mindustry/input/Placement.java

```diff
@@ -56,6 +56,31 @@
         return points;
     }
 
+    public static Array<Point2> followLineUntil(int startX, int startY, int endX, int endY){
+        Pools.freeAll(points);
+        points.clear();
+
+        Tile cursor = world.ltile(startX, startY);
+        Block filter = cursor.block();
+
+        while(true){
+            if(cursor.block() != filter) break;                        // keep looking until encountering a different block
+            if(points.contains(new Point2(cursor.x, cursor.y))) break; // avoid biting its own tail
```

Comment:
Use a predicate, don't create new points for checks: `contains(p -> p.equals(cursor.x, cursor.y))`.